Implement visibility range and dependencies

JFonS opened the pull request:
This commit adds the following properties to GeometryInstance3D: `visibility_range_begin`,
`visibility_range_begin_margin`, `visibility_range_end`, `visibility_range_end_margin`.

Together they define a range in which the GeometryInstance3D will be visible from the camera,
taking hysteresis into account for state changes. A begin or end value of 0 will be ignored,
so the visibility range can be open-ended in both directions.

This commit also adds the `visibility_parent` property to 'Node3D'.
Which defines the visibility parents of the node and its subtree (until
another parent is defined).

Visual instances with a visibility parent will only be visible when the parent, and all of its
ancestors recursively, are hidden because they are closer to the camera than their respective
`visibility_range_begin` thresholds.

Combining visibility ranges and visibility parents users can set-up a quick HLOD system
that shows high detail meshes when close (i.e buildings, trees) and merged low detail meshes
for far away groups (i.e. cities, woods).

This is a simple example "city":

https://user-images.githubusercontent.com/4402304/118841427-99e7b000-b8c8-11eb-9bcd-53b7fe65c1df.mp4

And this is a stress test with 50k of such "cities" (900k meshes in total):

https://user-images.githubusercontent.com/4402304/118842300-5fcade00-b8c9-11eb-9bfe-ba4caeb5d527.mp4

*Bugsquad edit: This closes https://github.com/godotengine/godot-proposals/issues/692.*

JFonS committed:
Implement visibility range and dependencies.

This commit adds the following properties to GeometryInstance3D: `visibility_range_begin`,
`visibility_range_begin_margin`, `visibility_range_end`, `visibility_range_end_margin`.

Together they define a range in which the GeometryInstance3D will be visible from the camera,
taking hysteresis into account for state changes. A begin or end value of 0 will be ignored,
so the visibility range can be open-ended in both directions.

This commit also adds the `visibility_parent` property to 'Node3D'.
Which defines the visibility parents of the node and its subtree (until
another parent is defined).

Visual instances with a visibility parent will only be visible when the parent, and all of its
ancestors recursively, are hidden because they are closer to the camera than their respective
`visibility_range_begin` thresholds.

Combining visibility ranges and visibility parents users can set-up a quick HLOD system
that shows high detail meshes when close (i.e buildings, trees) and merged low detail meshes
for far away groups (i.e. cities, woods).

clayjohn commented on doc/classes/GeometryInstance3D.xml:
```diff
@@ -52,25 +52,27 @@
-			The GeometryInstance3D's min LOD distance.
-			[b]Note:[/b] This property currently has no effect.
+		<member name="visibility_range_begin_hysteriesis" type="float" setter="set_visibility_range_begin_hysteriesis" getter="get_visibility_range_begin_hysteriesis" default="0.0">
+			Hysteresis margin for the [member visibility_range_begin] threshold.
```
Maybe add another sentence about what hysteresis is used for. Something like "hysteresis is the amount of overlap between two visibility ranges allowing both HLODs to draw at the same time to reduce popping in and out"

JFonS replied:
Right, I will add something like that :)

clayjohn commented on scene/3d/visual_instance_3d.cpp:
```diff
@@ -150,43 +150,87 @@ Ref<Material> GeometryInstance3D::get_material_override() const {
 
-float GeometryInstance3D::get_lod_max_hysteresis() const {
-	return lod_max_hysteresis;
+float GeometryInstance3D::get_visibility_range_begin_hysteriesis() const {
```
typo, should be "hysteresis", this is spelled wrong throughout

clayjohn reviewed (commented):
Looks really good on a quick review. You'll need to do a search and replace to replace all instances of "hysteriesis" with "hysteresis"

fire commented:
Are you using the LOD metric that is used by mesh optimizer?

JFonS commented:
> Looks really good on a quick review. You'll need to do a search and replace to replace all instances of "hysteriesis" with "hysteresis"
Oops, will do :P

> Are you using the LOD metric that is used by mesh optimizer?
No, AFAIK the mesh LOD metric uses the edge lengths and this system has to work on all GeometryInstance3D types. Also, a simple distance  check with hysteresis should work just fine for this kind of HLOD.

Calinou commented:
Does this visibility range system take the camera FOV into account? I wonder how we can keep using ranges while supporting wider/narrower camera FOVs.

clayjohn reviewed (commented):
Looks good to me!

reduz commented:
I am thinking that as much as it pains me, we should probably change it to a system where the parent has a visibility override over the children and not backwards, else usability will be poor.
This also means storing a list of children in the node, which will make editing more difficult and it may require a special editor. I can work on it I suppose, though I am just unhappy.

JFonS commented:
I updated the PR with the following changes:

* `visibility_range_begin` and `visibility_range_end` are disabled if their value is 0. This means I was able to remove the previous `visibility_range_enabled` property, and visibility checks are internally enabled only when any of the range values is not 0.
* I changed the visibility_parent behavior so that it only shows it's "children" instances when the camera is too close, while before it would show children whenever the instance was outside of the visibility range (both too far and too close).  This means most LOD levels don't have to check for a "range_end" threshold anymore, making the system more user-friendly and faster.

reduz commented on scene/3d/visual_instance_3d.cpp:
```diff
@@ -357,17 +440,20 @@ void GeometryInstance3D::_bind_methods() {
 
-	ClassDB::bind_method(D_METHOD("set_lod_min_distance", "mode"), &GeometryInstance3D::set_lod_min_distance);
-	ClassDB::bind_method(D_METHOD("get_lod_min_distance"), &GeometryInstance3D::get_lod_min_distance);
+	ClassDB::bind_method(D_METHOD("set_visibility_range_parent", "parent"), &GeometryInstance3D::set_visibility_range_parent);
```
I would most likely do this internally, and not expose the parent range or clearing parent to the user. As in, user should only need to be able to set the children.

JFonS replied:
What do you mean with "set the children"? Exposing a method that receives a list of nodes and then sets their parent internally?

The issue with that is it hides the fact that nodes can only have one parent, if properly documented it won't be that bad, but I'm still not sure what are the benefits of setting the children instead of setting the parent...

I can see why it's needed in the editor, where you just edit one node and select all of its children, but from the scripting side it's just a matter of iterating your list of children and setting their parent node.

JFonS replied:
Fixed in the latest version of the PR.

reduz commented on scene/3d/visual_instance_3d.cpp:
```diff
@@ -150,43 +150,89 @@ Ref<Material> GeometryInstance3D::get_material_override() const {
 }
 
 void GeometryInstance3D::_notification(int p_what) {
+	if (p_what == NOTIFICATION_ENTER_TREE) {
```
should probably use NOTIFICATION_READY to make sure children are properly configured (and call request_ready() afterwards so its called again next time it enters the tree.)

JFonS replied:
I don't think this is needed anymore. Right now, the only requirement is for the node to be inside a tree. If any nodes are added after that, they will automatically inherit the correct visibility parent.

JFonS commented:
Pushed the latest version with a new way to define visibility dependencies and fixes for all the mentioned issues.

reduz commented:
Looks fantastic, really great job with this!

JFonS commented:
I would wait for #48207 to be merged before merging this one. I will take care of the merge conflicts then

JFonS commented:
@akien-mga Rebased!

akien-mga commented:
Thanks!
